backends/lean/Aeneas/Progress/ProgressStar.lean

```diff
@@ -0,0 +1,326 @@
+  toMessageData discr := 
+    let nameMD := if let some name := discr.name? then m!"(name {name }) " else ""
+    m!"(discr {nameMD}{discr.toExpr})"
+
+/-- One possible branch in a bifurcation -/
+structure Branch where
+  /-- The branch as a function -/
+  toExpr: Expr
+
+  /-- The number of arguments the bifurcation is 
+      expected to provide. -/
+  numArgs: Nat
+
+  deriving Repr
+instance: ToMessageData Branch where
+  toMessageData br := m!"(branch (numArgs {br.numArgs}){br.toExpr}) "
+
+/-- The kind of bifurcation -/
+inductive Kind where
+  /-- An `if-then-else` bifurcation -/
+  | ite
+  /-- A dependent `if-then-else` bifurcation -/
+  | dite
+  /-- A matcher or cases bifurcation, of name `name` -/
+  | matcher (name: Name)
+  deriving Repr
+instance: ToString Kind where
+  toString
+  | .ite => "ite"
+  | .dite => "dite"
+  | .matcher name => s!"matcher {name}"
+instance: ToMessageData Kind where toMessageData k := toString k
+
+/-- Rough equivalent of `Lean.Meta.MatcherApp`, but which also includes if-then-else -/
+structure Info where
+  /-- The kind of bifurcation -/
+  kind: Kind
+
+  /-- The information on discriminants of the bifurcation -/
+  discrs: Array Discr
+
+  /-- The information on the branches of the bifurcation -/
+  branches: Array Branch
+
+  /-- The name of the function which implements the matcher -/
+  matcher: Name
+
+  /-- The motive of the bifurcation -/
+  motive: Expr
+
+  /-- The universe levels of the bifurcation -/
+  uLevels: List Lean.Level
+
+  /-- -/
+  params: Array Expr
+  deriving Repr
+instance: ToMessageData Info where
+  toMessageData 
+  | {kind, discrs, branches, ..} =>
+    let discr := MessageData.ofArray <| discrs.map (ToMessageData.toMessageData)
+    let branches := MessageData.ofArray <| branches.map (ToMessageData.toMessageData)
+    m!"(info {kind} {discr} {branches})"
+
+def Info.ofExpr(e: Expr): MetaM (Option Info) := do
+  let e := e.consumeMData
+  if e.isIte || e.isDIte then
+    let kind := if e.isIte then .ite else .dite
+    --   ite.{u} {α : Sort u} (c : Prop) [h : Decidable c] (t           e : α)      : α
+    --  dite.{u} {α : Sort u} (c : Prop) [h : Decidable c] (t : c → α) (e : ¬c → α) : α
+    /- let #[_ty, cond, _dec, brThen, brElse] := e.getAppArgs -/
+    /- let e ← whnf e -/ 
+    let e ← deltaExpand e (fun n => n == ``ite || n == ``dite)
+    -- Decidable.casesOn.{u} {prop} {motive} dec (isFalse: (h:¬p) → motive (isFalse h)) (isTrue: (h:p) → motive (isTrue h)) : motive t
+    let .const ``Decidable.casesOn uLevels := e.getAppFn
+      | throwError "Expected ``Decidable.rec, found {←ppExpr e.getAppFn}"
+    let #[prop, motive, dec, brFalse, brTrue] := e.getAppArgs
+      | throwError "Wrong number of parameters for {e.getAppFn}: {e.getAppArgs.size} [{e.getAppArgs}]"
+    return some {
+      kind,
+      discrs := #[{ toExpr := dec }]
+      -- TODO: I should be able to retrieve the name given to
+      --  the condition of a dite.
```

Comment:
I think the way to do this is to destruct the lambda of one of the branches and get the name given to the bound variable